src/main/java/com/pingcap/tikv/statistics/Histogram.java

```diff
@@ -0,0 +1,209 @@
+    // where table_id = %d and is_index = %d and hist_id = %d", tableID, isIndex, colID
+        List<TiExpr> firstAnd = ImmutableList.of(
+                new Equal(TiColumnRef.create("table_id", table), TiConstant.create(tableID)),
+                new Equal(TiColumnRef.create("is_index", table), TiConstant.create(isIndex)),
+                new Equal(TiColumnRef.create("hist_id", table), TiConstant.create(colID))
+        );
+
+        ScanBuilder scanBuilder = new ScanBuilder();
+        ScanBuilder.ScanPlan scanPlan = scanBuilder.buildScan(firstAnd, index, table);
+
+        TiSelectRequest selReq = new TiSelectRequest();
+        selReq.addRanges(scanPlan.getKeyRanges())
+                .setTableInfo(table)
+                .setIndexInfo(index)
+                .addField(TiColumnRef.create("table_id", table))
+                .addField(TiColumnRef.create("is_index", table))
+                .addField(TiColumnRef.create("hist_id", table))
+                .addField(TiColumnRef.create("bucket_id", table))
+                .addField(TiColumnRef.create("count", table))
+                .addField(TiColumnRef.create("repeats", table))
+                .addField(TiColumnRef.create("lower_bound", table))
+                .addField(TiColumnRef.create("upper_bound",table))
+                .setStartTs(snapshot.getVersion());
+
+        if (conf.isIgnoreTruncate()) {
+            selReq.setTruncateMode(TiSelectRequest.TruncateMode.IgnoreTruncation);
+        } else if (conf.isTruncateAsWarning()) {
+            selReq.setTruncateMode(TiSelectRequest.TruncateMode.TruncationAsWarning);
+        }
+
+        selReq.addWhere(PredicateUtils.mergeCNFExpressions(scanPlan.getFilters()));
+
+
+        List<RangeSplitter.RegionTask> keyWithRegionTasks =
+                RangeSplitter.newSplitter(cluster.getRegionManager())
+                        .splitRangeByRegion(selReq.getRanges());
+        for (RangeSplitter.RegionTask worker : keyWithRegionTasks) {
+            Iterator<Row> it = snapshot.select(selReq, worker);
+
+            while (it.hasNext()) {
+                Row row = it.next();
+                SchemaInfer schemaInfer = SchemaInfer.create(selReq);
+                long buckID = row.getLong(0);
+                long count = row.getLong(1);
+                long repeats = row.getLong(2);
+                if(isIndex == 1) {
+                    Bucket.lowerBound = Comparables.wrap(row.getLong(3));
+                }else {
+                    Bucket.lowerBound = (Comparable<ByteString>)
+                            row.get(row.getInteger(3), DataTypeFactory.of(Types.TYPE_BLOB));
+                    Bucket.upperBound = (Comparable<ByteString>)
+                            row.get(row.getInteger(4), DataTypeFactory.of(Types.TYPE_BLOB));
+                }
+                for (int i = 0; i < row.fieldCount(); i++) {
+                    Object val = row.get(i, schemaInfer.getType(i));
```

Comment:
why do you do this? the value of row has been extracted already. In addition, I did not see any where that you are creating a histogram. As the method name suggests, this method will retrieve a histogram from `TiKV`. But you did not save such information.